data-prepper-plugins/sns-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/SNSSinkService.java

```diff
@@ -0,0 +1,215 @@
+        }
+        reentrantLock.unlock();
+    }
+
+    private void processRecords() throws InterruptedException {
+        final AtomicReference<String> errorMsgObj = new AtomicReference<>();
+        final boolean isFlushToSNS = retryFlushToSNS(processRecordsList, topicName,errorMsgObj);
+        if (isFlushToSNS) {
+            numberOfRecordsSuccessCounter.increment(processRecordsList.size());
+        } else {
+            numberOfRecordsFailedCounter.increment(processRecordsList.size());
+            dlqPushHandler.perform(pluginSetting,new SNSSinkFailedDlqData(topicName,errorMsgObj.get(),0));
+        }
+        releaseEventHandles(true);
+    }
+
+    private void releaseEventHandles(final boolean result) {
+        for (EventHandle eventHandle : bufferedEventHandles) {
+            eventHandle.release(result);
+        }
+        bufferedEventHandles.clear();
+    }
+
+    /**
+     * perform retry in-case any issue occurred, based on max_upload_retries configuration.
+     *
+     * @param processRecordsList records to be process
+     * @param errorMsgObj pass the error message
+     * @return boolean based on object upload status.
+     * @throws InterruptedException interruption during sleep.
+     */
+    protected boolean retryFlushToSNS(final List<String> processRecordsList,
+                                      final String topicName,
+                                      final AtomicReference<String> errorMsgObj) throws InterruptedException {
+        boolean isUploadedToSNS = Boolean.FALSE;
+        int retryCount = maxRetries;
+        do {
+            try {
+                publishToTopic(snsClient, topicName,processRecordsList);
+                isUploadedToSNS = Boolean.TRUE;
+            } catch (AwsServiceException | SdkClientException e) {
+                errorMsgObj.set(e.getMessage());
+                LOG.error("Exception occurred while uploading records to sns. Retry countdown  : {} | exception:",
+                        retryCount, e);
+                --retryCount;
+                if (retryCount == 0) {
+                    return isUploadedToSNS;
+                }
+                Thread.sleep(5000);
+            }
+        } while (!isUploadedToSNS);
+        return isUploadedToSNS;
+    }
+
+    public void publishToTopic(SnsClient snsClient, String topicName,final List<String> processRecordsList) {
+        LOG.debug("Trying to Push Msg to SNS: {}",topicName);
+        try {
+            final PublishBatchRequest request = createPublicRequestByTopic(topicName, processRecordsList);
+            final PublishBatchResponse publishBatchResponse = snsClient.publishBatch(request);
+            LOG.info(" Message sent. Status is " + publishBatchResponse.sdkHttpResponse().statusCode());
+        }catch (Exception e) {
+            throw e;
+        }
+    }
+
+    private PublishBatchRequest createPublicRequestByTopic(final String topicName, final List<String>  processRecordsList) {
```

Comment:
This method will throw an SDK exception if topic is invalid for instance. We should have a clear log message on line 193 based on the exception.

Reply:
Thanks @asifsmohammed for your feedback, Sure, it will available in next commit for review.